### PR TITLE
fix(typings): Add combineEpics overload for custom Epic signatures

### PR DESCRIPTION
**Pull request by inuscript**
- [x] I have read the [Contributor Guide](https://github.com/redux-observable/redux-observable/blob/master/CONTRIBUTING.md)
- [x] My commit messages are in [conventional-changelog-standard](https://github.com/redux-observable/redux-observable/blob/master/CONTRIBUTING.md#sending-a-pull-request) format.

I try to use API injection epic pattern that like https://github.com/redux-observable/redux-observable/issues/84#issuecomment-236042600 and #163.

But current typing is not accept `combineEpics` with extra argument epics.

### CODE REVIEW

**Comment by jayphelps** [comment]
hmmm this would lost type safety on those extra args since they're all `any`.

What about instead creating an overload for `combineEpics` which allows you pass in **your own** implementation of the Epic interface, so you can have safety.

```ts
declare function combineEpics<T, S>(...epics: Epic<T, S>[]): Epic<T, S>;
// new overload
declare function combineEpics<T>(...epics: T[]): T;
```

```ts
interface CustomEpic {
  (action$: ActionsObservable<Action>, store: MiddlewareAPI<State>, api: SomeApi): Observable<T>;
}

const epic1: CustomEpic = (action$, store, api: SomeApi) =>
  action$.ofType('STUFF').etc();

const rootEpic: Epic<Action, State> = (action$, store) =>
  combineEpics<CustomEpic>(epic1)(action$, store, api);
```

**Comment by inuscript** [comment]
@jayphelps
Thank you for good advice. I change to this.

**Comment by jayphelps** [comment]
Seems good to me. Wonder if @kwonoj could spare a second to confirm this solution is kosher?

File: index.d.ts [hunk 1]
```diff
@@ -45,5 +45,7 @@ interface Options {
 }
 
 export declare function createEpicMiddleware<T, S>(rootEpic: Epic<T, S>, options?: Options): EpicMiddleware<T, S>;
+export declare function createEpicMiddleware<E, T, S>(rootEpic: E, options?: Options): EpicMiddleware<T, S>;
```
Comment (jayphelps) [inline index.d.ts]: Is this overload necessary? Shouldn't the rootEpic always be `(action$, store)` only?

Comment (inuscript) [inline index.d.ts]: Thank you. I didn't know this.
https://github.com/redux-observable/redux-observable/blob/master/src/createEpicMiddleware.js#L34

It's not good overload. I removed it.

**Comment by kwonoj** [comment]
This seems bits complicated to me, so this is based on assumption custom epics are following `<E>` always, is that correct? Though it's general problem of generic that can't carry multiple types easily, if there's non-formed type of epic than it won't be accepted. Like, even default epic definition falls into those category like

```js
const defaultEpic = (a, s) => {};
const customEpic: CustomEpic<U> = (...) => {};

combineEpics<CustomEpic<U>>(defaultEpic, customEpic); //can't apply both epics?
```

**Comment by jayphelps** [comment]
@kwonoj hmm I'm not 100% sure what you mean. This new overload is so that people can provide a custom signature for epics--usually because they're doing dependency injection. All of their epics  have to have the same signature.

**Comment by kwonoj** [comment]
Yes, my initial assumption was someone would expect mixed types between different types of Epics. If this overload is intended to be used for constraint to epic have all same custom interfaces, I think it'll be fine.

**Comment by jayphelps** [comment]
@kwonoj Someone could presumably provide a union of multiple signatures as well AFAICT

**Comment by jayphelps** [comment]
Thanks @inuscript ! 🎉

**Pull request merged**